### ISSUE TITLE
If I will want rewrite service (services.Service) on another language, what can I do?

### ISSUE THREAD

**Issue by viatoriche** [zatosource/zato#548]
Let I have service with slow implemetation, and I do not want write this on cython.
Let I want write my service on Go (miscroservice arch.)
What can I do for this? Have a Zato API for external languages, adapters? Or I need write this adapter for use my service in zato system?

**Comment by dsuch** [comment]
What is the service in Python that has a slow implementation? Can you please post its source code? What does the service do? Why is it slow?

Thanks!

**Comment by viatoriche** [comment]
@dsuch, I can not this service, I just think about future.
Example: work with big requests, long-calculation, serialization/deserialization big-data, big xml, bad protobuf objects (many repeated elements). This is just discussion, not real problem. I just choosing technologies for my esb service and thinking about extendability

**Comment by dsuch** [comment]
Hi @viatoriche - you need to keep in mind that most of JSON/XML parsers in Python world are written in C. You don't parse anything in Python. In other words, there is not much to slow you down.

lxml for instance is a fairly thin Cython-based wrapper around libxml - check out its source code, it's very readable and you'll see this is all C with a thin layer of Cython on top of it https://github.com/lxml/lxml

As far as JSON goes - parsers are fast and the only thing you need to think of is RAM usage. In my experience, with documents I am working with, pyrapidjson is slightly faster than ujson and uses significantly less RAM.

I don't work with protobuf currently so I cannot comment on that.

In other words - all the heavy processing is actually done in C (or C++ in case of pyrapidjson). There is not much to speed up without actually programming in C or C++.

That all said - if you need to implement anything in Go, I haven't done it personally but I can see that Go programs can be compiled to shared libraries, so you should be able to simply import such a library in your Python service and call it:

http://www.darkcoding.net/software/building-shared-libraries-in-go-part-1/

**Comment by viatoriche** [comment]
@dsuch thx for your answer. Is there plans for create simple API's for integration any services, which wrote on no-python languages? Support microservices paradigma?

**Comment by dsuch** [comment]
@viatoriche - what we do in Zato is already microservices architecture and in fact we had been doing it well before the term was coined. 

As for calling non-Python code, I don't see why not. But let's also remember that there are already APIs in Python for invoking shared libraries:

* ctypes
* cffi

So anything on top of them would be rather thin. In other words, yes, why not if it turns out it is useful but unless you are absolutely sure you need it then you should simply not worry about it - Python itself is very fast and there are many fast C libraries with Python wrappers.

**Comment by viatoriche** [comment]
@dsuch I know about ctypes, cython, cffi. I mean network API (REST or RPC), for integrate isolated microservices into the zato system.

**Comment by dsuch** [comment]
Hi @viatoriche - can you please describe in details what you mean by "integrating isolated microservices" into Zato?

Integrating other APIs is precisely what Zato does so I'm not 100% sure what you mean? Can you please say more about your processes and use-cases?

Thanks a lot!

**Comment by viatoriche** [comment]
Let: I have some service (services.Service)

This service works ok.

But some requirements or libs for service was out of date (or something happened and we cannot use this service)

We want rewrite service on another language, with another requirements. And/Or we want move service on cloud, for example: lambda function on Amazon Web Services.

What we can do? How integrate lambda's services to the zato. I see one solution now: write service-adapter for outer service, but may be better create network API, where service can register as native zato service without adapter?

**Comment by dsuch** [comment]
I see your point and the idea with registrations sounds OK.

We don't have it right now in Zato but it works in a very similar manner already.

Let's say you have an outgoing REST connection called 'get.customer.data'.

https://zato.io/docs/web-admin/outgoing/plain-http.html

In your services you just refer to it by its name 'get.customer.data', i.e. you don't explicitly connect to any TCP/REST endpoint yourself:

https://zato.io/docs/progguide/rest/outconns.html

Now, should that API become unavailable for any reason you can just update the endpoint's definition in web-admin and make it point to a new address instead of the previous one. At that point your service will work just like that automatically - it will use the new endpoint without any coding required.

Another point is that if you use REST publish/subscribe you already have a registry of subscribers https://zato.io/docs/pubsub/index.html so should one of them become unavailable you can just start a new one, subscribe to a topic and no programming changes will be needed. That new one can be in Python, Go, Java, C# or anything.

**Comment by viatoriche** [comment]
@dsuch Thank you!
